File: c-sharp/Aoc.2024/Day1.cs
using Aoc.Shared.Abstract;

namespace Aoc._2024;

public class Day1 : IAocSolution
{
    public int Day => 1;

    private (int[], int[]) GetValues()
    {
        var filePaths = new FilePaths();
        var values = File.ReadAllLines(filePaths.InputDataPath(Day)).Select(line => line.Split(' ', StringSplitOptions.RemoveEmptyEntries)).ToArray();
        var lefts = values.Select(pair => int.Parse(pair[0])).Order().ToArray();
        var rights = values.Select(pair => int.Parse(pair[1])).Order().ToArray();
        
        return (lefts, rights);
    }

    public string Part1()
    {
        var (lefts, rights) = GetValues();
        int result = lefts.Zip(rights, (left, right) => Math.Abs(left - right)).Sum();

        return result.ToString();
    }

    public string Part2()
    {
        var (lefts, rights) = GetValues();
        var result = lefts.Sum(left => left * rights.Count(right => right == left));
        return result.ToString();
    }

}

File: python/day_1.py
from pathlib import Path

SAMPLE_PATH = Path(__file__).parent.parent / "sample" / "day_1.txt"
INPUT_PATH = Path(__file__).parent.parent / "input" / "day_1.txt"


def part_1() -> int:
    with INPUT_PATH.open("r") as f:
        data = [line.split() for line in f.readlines()]
        left, right = list(zip(*data))

    return sum(abs(int(l) - int(r)) for l, r in zip(sorted(left), sorted(right)))


def part_2() -> int:
    with INPUT_PATH.open("r") as f:
        data = [line.split() for line in f.readlines()]
        left, right = list(zip(*data))
        return sum(int(l) * right.count(l) for l in left)


if __name__ == '__main__':
    print(part_2())
